### ISSUE TITLE
ID's from generate_batch and forward_batch don't agree

### ISSUE THREAD

**Issue by RockyZhu29** [OpenNMT/CTranslate2#1477]
For text generation with greedy decoding, we first obtain sequence_ids generated using generate_batch [1]. Then, the sequence_ids are parsed into forward_batch and argmax are taken using torch.tensor(np.array(model.forward_batch(sequence_ids))).argmax(dim = -1) [2]. One would expect the sequences_ids that correspond to the generated texts to agree in [1] and [2], but in reality they are quite different. I wonder why is this the case?

**Comment by guillaumekln** [comment]
Can you post a complete code snippet to reproduce the issue?

**Comment by guillaumekln** [comment]
Thank you for the code snippet. There are 2 things happening here:

First, note that the tokenized prompt starts with the special token `</s>` which is not included in the generated tokens. You should add this token back before calling `forward_batch`, e.g.:

```python
generator.forward_batch([[tokenizer.bos_token_id] + out[0].sequences_ids[0]])
```

Second, there are known numerical issues for OPT models in int8 (see #818). As mentioned in the [OPT example](https://opennmt.net/CTranslate2/guides/transformers.html#opt), it is recommended to download the corresponding activation scales and pass them to the conversion command, e.g.:

```
ct2-transformers-converter --model facebook/opt-1.3b --output_dir opt-1.3b --quantization int8 --activation_scales opt-1.3b.pt
```

When you applied these 2 changes, you finally get the same IDs:

```
Generated new tokens:  tensor([100, 437,  45, 686,   4,  38, 437,  45, 686,  38])
Forward new tokens:  tensor([[100, 437,  45, 686,   4,  38, 437,  45, 686,  38]])
```

**Comment by RockyZhu29** [comment]
Thank you for your fast response. I updated my code accordingly, but the problem still persists:

 ``` 
import ctranslate2
import subprocess
from transformers import AutoTokenizer
import numpy as np
import torch

model = "facebook/opt-1.3b"
optimized_model_path = "ct2_model"

ct2_command = [
    "ct2-transformers-converter",
    "--model",
    model,
    "--quantization",
    "int8",
    "--output_dir",
    optimized_model_path,
    #"--trust_remote_code",
    #"--low_cpu_mem_usage",
    #'--force',
    "--activation_scales",
    "opt-1.3b.pt"
] 

subprocess.run(ct2_command)
 ``` 
 ``` 
# load the model 
tokenizer = AutoTokenizer.from_pretrained(model)
generator = ctranslate2.Generator(optimized_model_path, device="cpu") # 'cuda' for GPU

# prepare for inference 
prompt = "Hey, are you conscious? Can you talk to me?"
input_ids = tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt))

# -- generate 10 new tokens -- 

# generate using 'generate_batch'
out = generator.generate_batch([input_ids],max_length = len(input_ids) + 10, sampling_topk=1)
generate_new_tokens = torch.tensor(out[0].sequences_ids[0][-10:])

# generate using 'forward_batch', here we use argmax as greedy decoding 
forward = torch.tensor(np.array(generator.forward_batch([[tokenizer.bos_token_id]+out[0].sequences_ids[0]],return_log_probs = True)))[:, -10 -1:, :]
forward_new_tokens = forward.argmax(dim = -1)[:,:-1]

print("Generated new tokens: " , generate_new_tokens) # tensor([50118,   133,  1948,    16,  4420,     6,    53,    24,    17,    27])
print("Forward new tokens: " , forward_new_tokens) # tensor([[50118,   100,   1948,    16,  4420,     6,    53,    24,    17,    27]])

# Now the second token is different! 
 ``` 
Could you please let me know what I have misunderstood, thanks!

**Comment by guillaumekln** [comment]
Most likely the conversion command silently failed because `ct2_model` already exists on disk. Uncomment the option `'--force'` and try again.
